test/eme.test.js

```diff
@@ -1,9 +1,132 @@
 import QUnit from 'qunit';
 import videojs from 'video.js';
-import {standard5July2016} from '../src/eme';
+import {
+  standard5July2016,
+  makeNewRequest
+} from '../src/eme';
 
 QUnit.module('videojs-contrib-eme eme');
 
+QUnit.test('keystatuseschange with expired key closes session', function(assert) {
+  const listeners = [];
+  let numCloses = 0;
+  // mock session to make testing easier (so we can trigger events)
+  const mockSession = {
+    addEventListener: (type, listener) => listeners.push({ type, listener }),
+    generateRequest(initDataType, initData) {
+      // noop
+      return new Promise((resolve, reject) => resolve());
+    },
+    keyStatuses: new Map(),
+    close: () => {
+      numCloses++;
+      // fake a promise for easy testing
+      return {
+        then: (nextCall) => nextCall()
+      };
+    }
+  };
+  const removeSessionCalls = [];
+  // once the eme module gets the removeSession function, the session argument is already
+  // bound to the function (note that it's a custom session maintained by the plugin, not
+  // the native session), so only initData is passed
+  const removeSession = (initData) => removeSessionCalls.push(initData);
+  const initData = new Uint8Array([1, 2, 3]);
+
+  makeNewRequest({
+    mediaKeys: {
+      createSession: () => mockSession
+    },
+    initDataType: '',
+    initData,
+    options: {},
+    getLicense() {},
+    removeSession
+  });
+
+  assert.equal(listeners.length, 2, 'added listeners');
+  assert.equal(listeners[1].type,
+               'keystatuseschange',
+               'added keystatuseschange listener');
+  assert.equal(numCloses, 0, 'no session close calls');
+  assert.equal(removeSessionCalls.length, 0, 'no removeSession calls');
+
+  // no key statuses
+  listeners[1].listener();
+
+  assert.equal(numCloses, 0, 'no session close calls');
+  assert.equal(removeSessionCalls.length, 0, 'no removeSession calls');
+
+  mockSession.keyStatuses.set(1, 'unrecognized');
+  listeners[1].listener();
+
+  assert.equal(numCloses, 0, 'no session close calls');
+  assert.equal(removeSessionCalls.length, 0, 'no removeSession calls');
+
+  mockSession.keyStatuses.set(2, 'expired');
+  listeners[1].listener();
+
+  assert.equal(numCloses, 1, 'closed session');
+  // close promise is fake and resolves synchronously, so we can assert removes
+  // synchronously
+  assert.equal(removeSessionCalls.length, 1, 'called remove session');
+  assert.equal(removeSessionCalls[0], initData, 'called to remove session with initData');
+});
+
+QUnit.test('keystatuseschange with internal-error logs a warning', function(assert) {
+  const origWarn = videojs.log.warn;
+  const listeners = [];
+  // mock session to make testing easier (so we can trigger events)
+  const mockSession = {
```

Comment:
You could have one mockSession object and share it between the two tests, I don't think the additional close function above would negatively effect this test.

Reply:
That's a good point, and could save on some repeated code. I was kinda going by the rule that if we encounter a third case it should be shared (just cause I'm hesitant to share between tests with just two copied parts), but I can still change if you think we should.

Reply:
I agree with that rule, but I think that it might get dropped on the floor if we add another test here in the future (the reviewer might not spot the repetition), so it might be worth just adding it now to prevent that from happening.